courseme/main/views.py

```diff
@@ -141,128 +142,37 @@ def objectives_group(group_id, scheme_id=0, name_display=1):
-            #Check whether the name has changed and if so check it is valid
-            if proceed:
-                if name != objective.name:
-                    #Name has changed - need to check if new name already exists
-                    check_obj = g.user.visible_objectives().filter_by(
-                        name=name).first()  #DJG - code repeat of above, how to avoid this
-                    if check_obj is not None:
-                        #The new objective name is already taken
-                        result['edit_objective_name'] = ["Objective '" + name + "' already exists"]
-                        proceed = False
-
-            #Name not changed or new name not taken            
-            if proceed:
-                #Need to check user.subject is the same as the existing objective subject
-                if g.user.subject_id != objective.subject_id or (topic and topic.subject_id != objective.subject_id):
-                    result['edit_objective_subject'] = [objective.subject_id]
-                    proceed = False
+        if data['topic_id'] == u'0':
+            data['topic_id'] = None
 
-            #User subject same as exsting objective subject
-            if proceed:
-                #Need to check all the prerequisites exist already            
-                if undefined_prerequisites:  #DJG - code repeat of above, how to avoid this
-                    is_are = 'is not already defined as an objective' if len(
-                        undefined_prerequisites) == 1 else 'are not already defined as objectives'
-                    result['new_prerequisite'] = ["'" + "', '".join(undefined_prerequisites) + "' " + is_are]
-                else:
-                    #Need to check for cyclic prerequisites
-                    cyclic_prerequisites = [p.name for p in prerequisites if p.is_required_indirect(objective)]
-                    if cyclic_prerequisites:
-                        is_are = 'is' if len(cyclic_prerequisites) == 1 else 'are'
-                        result['new_prerequisite'] = ["'" + "', '".join(
-                            cyclic_prerequisites) + "' " + is_are + " dependent on the current objective"]
-                    else:
-                        objective.name = name
-                        objective.prerequisites = prerequisites
-                        objective.topic_id = topic_id
-                        db.session.add(objective)
-                        db.session.commit()
-                        result['savedsuccess'] = True
+        try:
+            if not data['id']:
+                data['subject_id'] = g.user.subject_id
+                service_layer.objectives.create(data, g.user)
+                return _ajax_success()
+            else:
+                service_layer.objectives.update(data, g.user)
+                return _ajax_success()
+
+        except ValidationError, e:
+            return _ajax_failure(**e.errors)
+        except NotAuthorised, e:
```

Comment:
Standard boiler-plate, should be handled by registering some sort of error handler with Flask.